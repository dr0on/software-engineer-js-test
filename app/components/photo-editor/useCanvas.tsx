import { useState, useEffect, useRef } from 'react';

export const enum ImageAction {
  Top,
  Down,
  Left,
  Right,
  ZoomIn,
  ZoomOut,
}

const useCanvas = () => {
  const [image, setImage] = useState<HTMLImageElement>(new Image());
  const [options, setOptions] = useState({ x: 0, y: 0, scale: 1 });
  const canvasRef = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    if (image && canvasRef.current) {
      const canvas = canvasRef.current;
      const ctx = canvas.getContext('2d');

      ctx?.clearRect(0, 0, canvas.width, canvas.height);
      ctx?.drawImage(
        image,
        options.x,
        options.y,
        image.naturalWidth * options.scale,
        image.naturalHeight * options.scale
      );
    }
  }, [image, options.x, options.y, options.scale]);

  const calculateZoomOptions = (img: any, scale: number) => {
    const currentWidth = img.width * options.scale;
    const currentHeight = img.height * options.scale;

    const newWidth = currentWidth * scale;
    const newHeight = currentHeight * scale;

    const xOffset = (newWidth - currentWidth) / 2;
    const yOffset = (newHeight - currentHeight) / 2;

    return {
      x: options.x - xOffset,
      y: options.y - yOffset,
      scale: options.scale * scale,
    };
  };

  const isBottomLimit = () => {
    const canvasHeight = canvasRef.current?.height || 0;
    const scaledHeight = image.height * options.scale;

    return options.y + scaledHeight - 10 < canvasHeight;
  };

  const isLeftLimit = () => {
    const canvasWidth = canvasRef.current?.width || 0;
    const scaledWidth = image.width * options.scale;

    return options.x + scaledWidth - 10 < canvasWidth;
  };

  const isTopLimit = () => options.y + 10 > 0;

  const isRightLimit = () => options.x + 10 > 0;

  const setZoomIn = () => {
    const { x, y, scale } = calculateZoomOptions(image, 1.1);

    setOptions({ x, y, scale });
  };

  const setZoomOut = () => {
    if (!isTopLimit() && !isBottomLimit() && !isLeftLimit() && !isRightLimit()) {
      const { x, y, scale } = calculateZoomOptions(image, 0.9);

      setOptions({ x, y, scale });
    }
  };

  const setMoveTop = () => {
    const y = isTopLimit() ? 0 : options.y + 10;
    setOptions({ x: options.x, y, scale: options.scale });
  };

  const setMoveBottom = () => {
    const canvasHeight = canvasRef.current?.height || 0;
    let y = isBottomLimit() ? canvasHeight - image.height * options.scale : options.y - 10;

    setOptions({ x: options.x, y, scale: options.scale });
  };

  const setMoveRight = () => {
    const x = isRightLimit() ? 0 : options.x + 10;

    setOptions({ x, y: options.y, scale: options.scale });
  };

  const setMoveLeft = () => {
    const canvasWidth = canvasRef.current?.width || 0;
    let x = isLeftLimit() ? canvasWidth - image.width * options.scale : options.x - 10;

    setOptions({ x, y: options.y, scale: options.scale });
  };

  const setImageAction = (action: ImageAction) => {
    switch (action) {
      case ImageAction.ZoomIn:
        setZoomIn();
        break;
      case ImageAction.ZoomOut:
        setZoomOut();
        break;
      case ImageAction.Top:
        setMoveTop();
        break;
      case ImageAction.Down:
        setMoveBottom();
        break;
      case ImageAction.Right:
        setMoveRight();
        break;
      case ImageAction.Left:
        setMoveLeft();
        break;
    }
  };

  return { canvasRef, image, setImage, options, setOptions, setImageAction };
};

export default useCanvas;
