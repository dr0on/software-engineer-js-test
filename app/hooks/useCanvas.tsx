import { useState, useEffect, useRef } from 'react';

export const enum ImageAction {
  Top,
  Down,
  Left,
  Right,
  ZoomIn,
  ZoomOut,
}

const MOVE_STEP = 10;

const useCanvas = () => {
  const [image, setImage] = useState<HTMLImageElement>(new Image());
  const [options, setOptions] = useState({ x: 0, y: 0, scale: 1 });
  const canvasRef = useRef<HTMLCanvasElement>(null);

  const canvasHeight = canvasRef.current?.height || 0;
  const canvasWidth = canvasRef.current?.width || 0;

  useEffect(() => {
    if (image && canvasRef.current) {
      const canvas = canvasRef.current;
      const ctx = canvas.getContext('2d');

      ctx?.clearRect(0, 0, canvas.width, canvas.height);
      ctx?.drawImage(
        image,
        options.x,
        options.y,
        image.naturalWidth * options.scale,
        image.naturalHeight * options.scale
      );
    }
  }, [image, options.x, options.y, options.scale]);

  const calculateZoomOptions = (img: any, scale: number) => {
    const currentWidth = img.width * options.scale;
    const currentHeight = img.height * options.scale;

    const newWidth = currentWidth * scale;
    const newHeight = currentHeight * scale;

    const xOffset = (newWidth - currentWidth) / 2;
    const yOffset = (newHeight - currentHeight) / 2;

    return {
      x: options.x - xOffset,
      y: options.y - yOffset,
      scale: options.scale * scale,
    };
  };

  const isBottomLimit = () => {
    const scaledHeight = image.height * options.scale;
    return options.y + scaledHeight - MOVE_STEP < canvasHeight;
  };

  const isLeftLimit = () => {
    const scaledWidth = image.width * options.scale;
    return options.x + scaledWidth - MOVE_STEP < canvasWidth;
  };

  const isTopLimit = () => options.y + MOVE_STEP > 0;

  const isRightLimit = () => options.x + MOVE_STEP > 0;

  const setZoomIn = () => {
    const { x, y, scale } = calculateZoomOptions(image, 1.1);
    setOptions({ x, y, scale });
  };

  const setZoomOut = () => {
    if (!isTopLimit() && !isBottomLimit() && !isLeftLimit() && !isRightLimit()) {
      const { x, y, scale } = calculateZoomOptions(image, 0.9);
      setOptions({ x, y, scale });
    }
  };

  const setMoveTop = () => {
    const y = isTopLimit() ? 0 : options.y + MOVE_STEP;
    setOptions({ x: options.x, y, scale: options.scale });
  };

  const setMoveBottom = () => {
    let y = isBottomLimit() ? canvasHeight - image.height * options.scale : options.y - MOVE_STEP;
    setOptions({ x: options.x, y, scale: options.scale });
  };

  const setMoveRight = () => {
    const x = isRightLimit() ? 0 : options.x + MOVE_STEP;
    setOptions({ x, y: options.y, scale: options.scale });
  };

  const setMoveLeft = () => {
    let x = isLeftLimit() ? canvasWidth - image.width * options.scale : options.x - MOVE_STEP;
    setOptions({ x, y: options.y, scale: options.scale });
  };

  const setImageAction = (action: ImageAction) => {
    switch (action) {
      case ImageAction.ZoomIn:
        setZoomIn();
        break;
      case ImageAction.ZoomOut:
        setZoomOut();
        break;
      case ImageAction.Top:
        setMoveTop();
        break;
      case ImageAction.Down:
        setMoveBottom();
        break;
      case ImageAction.Right:
        setMoveRight();
        break;
      case ImageAction.Left:
        setMoveLeft();
        break;
    }
  };

  return { canvasRef, image, setImage, options, setOptions, setImageAction };
};

export default useCanvas;
