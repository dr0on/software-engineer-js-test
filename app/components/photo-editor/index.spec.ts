/**
 * @jest-environment jsdom
 */

import React, { useEffect }  from 'react';
import { toDataURL } from './settingsUtils';
import 'jest-canvas-mock';

import useCanvas, { ImageAction } from './useCanvas';

let canvas: HTMLCanvasElement;
let ctx: CanvasRenderingContext2D | null;

// jest.mock('React', () => ({
//   ...jest.requireActual('React'),
//   useEffect: jest.fn(),
// }));

describe('PhotoEditor', () => {
  beforeEach(() => {
    canvas = document.createElement('canvas');
    ctx = canvas.getContext('2d');
    canvas.width = 400;
    canvas.height = 300;
  });

  it('should encode Image to base64', () => {
    const image = new Image(400, 300);    
    image.src = 'picture.jpg';

    const encodedImg = toDataURL(image);

    expect(encodedImg).toEqual('data:image/png;base64,00');
  });

  describe('useCanvas', () => {
    it('should return default options', () => {
      jest.spyOn(React, 'useState')
        .mockImplementationOnce(() => [new Image(), () => null])
        .mockImplementationOnce(() => [{ x: 0, y: 0, scale: 1 }, () => null]);
      jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: canvas });
      jest.spyOn(React, 'useEffect').mockImplementation(f => f());

      const { options } = useCanvas();

      expect(options).toEqual({ x: 0, y: 0, scale: 1 });
    });

    it('should return Ref to canvas', () => {
      jest.spyOn(React, 'useState')
        .mockImplementationOnce(() => [new Image(), () => null])
        .mockImplementationOnce(() => [{ x: 0, y: 0, scale: 1 }, () => null]);
      jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: canvas });
      jest.spyOn(React, 'useEffect').mockImplementation(f => f());

      const { canvasRef } = useCanvas();

      expect(canvasRef.current).toEqual(canvas);
    });

    it('should return image', () => {
      jest.spyOn(React, 'useState')
        .mockImplementationOnce(() => [new Image(200, 100), () => null])
        .mockImplementationOnce(() => [{ x: 0, y: 0, scale: 1 }, () => null]);
      jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: canvas });
      jest.spyOn(React, 'useEffect').mockImplementation(f => f());

      const { image } = useCanvas();

      expect(image.width).toEqual(200);
    });

    it('should zoom in image', () => {
      jest.spyOn(React, 'useState')
        .mockImplementationOnce(() => [new Image(200, 100), () => null])
        .mockImplementationOnce(() => [{ x: 0, y: 0, scale: 1 }, () => null]);
      jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: canvas });
      jest.spyOn(React, 'useEffect').mockImplementation(f => f());

      const { setImageAction } = useCanvas();

      setImageAction(ImageAction.ZoomIn);

      const events = ctx?.__getEvents();

      expect(events).toMatchSnapshot();
    });

    it('should move image to top', () => {
      jest.spyOn(React, 'useState')
        .mockImplementationOnce(() => [new Image(1000, 1000), () => null])
        .mockImplementationOnce(() => [{ x: 0, y: -15, scale: 1 }, () => null]);
      jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: canvas });
      jest.spyOn(React, 'useEffect').mockImplementation(f => f());

      const { setImageAction } = useCanvas();

      setImageAction(ImageAction.Top);

      const events = ctx?.__getEvents();

      expect(events).toMatchSnapshot();
    });
  });
});