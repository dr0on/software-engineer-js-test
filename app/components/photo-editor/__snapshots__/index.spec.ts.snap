// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoEditor useCanvas should move image to top 1`] = `
[
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 1000,
      "dWidth": 1000,
      "dx": 0,
      "dy": -15,
      "img": <img
        height="1000"
        width="1000"
      />,
      "sHeight": 1000,
      "sWidth": 1000,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`PhotoEditor useCanvas should zoom in image 1`] = `
[
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 100,
      "dWidth": 200,
      "dx": 0,
      "dy": 0,
      "img": <img
        height="100"
        width="200"
      />,
      "sHeight": 100,
      "sWidth": 200,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
